tokio/src/net/tcp/stream.rs

```diff
@@ -664,6 +665,65 @@ impl TcpStream {
         self.io.set_nodelay(nodelay)
     }
 
+    /// Reads the linger duration for this socket by getting the `SO_LINGER`
+    /// option.
+    ///
+    /// For more information about this option, see [`set_linger`].
+    ///
+    /// [`set_linger`]: TcpStream::set_linger
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::net::TcpStream;
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn std::error::Error>> {
+    /// let stream = TcpStream::connect("127.0.0.1:8080").await?;
+    ///
+    /// println!("{:?}", stream.linger()?);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn linger(&self) -> io::Result<Option<Duration>> {
+        let mio_socket = mio::net::TcpSocket::from(self);
```

Comment:
It should be possible to access `self.io.get_linger()`, similar to how `nodelay` does it above.

Reply:
Mio TcpStream doesn't have a linger option and that's because the `inner` uses std which also doesn't have linger options

Reply:
There are no linger options in either Mio's TcpStream which `self.io` refers to nor it's `inner` which is std::net::tcp

Reply:
@seanmonstar I actually don't think that's right; the methods for getting and setting SO_LINGER are defined on `mio`'s `TcpSocket` type, _not_ on its `TcpStream`, and `self.io` is a `TcpStream`.

That does bring up a different question, though: is there any reason it's necessary to expose these methods on Tokio's `TcpStream` type, rather than on its `TcpSocket` type? `tokio::net::TcpSocket` wraps `mio`'s `TcpSocket`...

Reply:
Oh, I'm sorry, I misread that as `TcpStream::from` instead of `TcpSocket`.

Reply:
@hawkw In our use case (pre tokio 0.3) we set the linger option from a stream object not TcpSocket.

I could also update this PR to also expose this on `tokio::net::TcpSocket`.

Reply:
I am a bit confused here. `set_linger` is on the mio socket and `PollEvented` holds the mio socket? `PollEvented` also derefs to the mio socket, so why does self.io.get_linger() not work?

Reply:
`PolledEvented` holds Mio `TcpStream`

Reply:
Ah yes, that is terribly confusing :laughing: